### ISSUE TITLE
FS-UAE crashes during startup when using SDL2 wayland backend

### ISSUE THREAD

**Issue by digitalstreamio** [FrodeSolheim/fs-uae#125]
It works fine w/ default SDL X11 fallback when running under Gnome Wayland Session (except for cursor locking issue w/ SDL). However, when using env var SDL_VIDEODRIVER=wayland, it crashes on startup.

*** Configuration: 
Fedora 24 w/ latest packages
Kernel 4.6.7-300.fc24.x86_64 
SDL 2.0.4 (8.fc24)
Mesa 12.0.1 (1.fc24)
FS-UAE 2.7.14dev3

*** Repro Steps: 
SDL_VIDEODRIVER=wayland fs-uae

*** Console Output:
FS-UAE 2.7.14dev3 (Built for Linux x86-64)
Copyright 1995-2002 Bernd Schmidt, 1999-2015 Toni Wilen,
2003-2007 Richard Drummond, 2006-2011 Mustafa 'GnoStiC' Tufan,
2011-2015 Frode Solheim, and contributors.

This is free software; see the file COPYING for copying conditions. There
is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE. See the README for more copyright info, and the source code for
a full list of contributors

UAE: Initializing core derived from WinUAE 3220
LOG: /home/xxx/FS-UAE/Cache/Logs/fs-uae.log.txt
Segmentation fault (core dumped)

*** Stacktrace:

,   "stacktrace":
      [ {   "crash_thread": true
        ,   "frames":
              [ {   "address": 140512243544349
                ,   "build_id": "318d2243f0c7c5428fb3bd6bd1e050ed90848c2f"
                ,   "build_id_offset": 229661
                ,   "function_name": "XQueryExtension"
                ,   "file_name": "/usr/lib64/libX11.so.6.3.0"
                }
              , {   "address": 140512243493554
                ,   "build_id": "318d2243f0c7c5428fb3bd6bd1e050ed90848c2f"
                ,   "build_id_offset": 178866
                ,   "function_name": "XInitExtension"
                ,   "file_name": "/usr/lib64/libX11.so.6.3.0"
                }
              , {   "address": 140512260378172
                ,   "build_id": "3b72fa30c16afaa6157f827eac5eb879a9d01746"
                ,   "build_id_offset": 122428
                ,   "function_name": "__glXInitialize"
                ,   "file_name": "/usr/lib64/libGL.so.1.2.0"
                }
              , {   "address": 140512260359665
                ,   "build_id": "3b72fa30c16afaa6157f827eac5eb879a9d01746"
                ,   "build_id_offset": 103921
                ,   "function_name": "glXQueryVersion"
                ,   "file_name": "/usr/lib64/libGL.so.1.2.0"
                }
              , {   "address": 140512265707668
                ,   "build_id": "cca9dfc42347857d2f1a6ecd513f401952cf57dd"
                ,   "build_id_offset": 336020
                ,   "function_name": "glxewInit"
                ,   "file_name": "/usr/lib64/libGLEW.so.1.13.0"
                }
              , {   "address": 12545969
                ,   "build_id": "fc5051aedaed87eb07e6b668bdbae9d746db0eff"
                ,   "build_id_offset": 8351665
                ,   "file_name": "/usr/local/bin/fs-uae"
                }
              , {   "address": 12473060
                ,   "build_id": "fc5051aedaed87eb07e6b668bdbae9d746db0eff"
                ,   "build_id_offset": 8278756
                ,   "file_name": "/usr/local/bin/fs-uae"
                }
              , {   "address": 4252945
                ,   "build_id": "fc5051aedaed87eb07e6b668bdbae9d746db0eff"
                ,   "build_id_offset": 58641
                ,   "file_name": "/usr/local/bin/fs-uae"
                } ]
        } ]
}

**Comment by digitalstreamio** [comment]
*** FS-UAE log
----------------------------------------------------------------------------
fs-uae init
----------------------------------------------------------------------------

configure logging: none
- checking /home/xxx/.config/fs-uae/kickstarts-dir
resolve_path /home/xxx/FS-UAE/Kickstarts (absolute)
- using "Kickstarts" directory "/home/xxx/FS-UAE/Kickstarts"
- checking /home/xxx/.config/fs-uae/floppies-dir
resolve_path /home/xxx/FS-UAE/Floppies (absolute)
- using "Floppies" directory "/home/xxx/FS-UAE/Floppies"
- checking /home/xxx/.config/fs-uae/cdroms-dir
resolve_path /home/xxx/FS-UAE/CD-ROMs (absolute)
- using "CD-ROMs" directory "/home/xxx/FS-UAE/CD-ROMs"
- checking /home/xxx/.config/fs-uae/hard-drives-dir
resolve_path /home/xxx/FS-UAE/Hard Drives (absolute)
- using "Hard Drives" directory "/home/xxx/FS-UAE/Hard Drives"
[PLUGINS] Initializing
- checking /home/xxx/.config/fs-uae/plugins-dir
resolve_path /home/xxx/FS-UAE/Plugins (absolute)
- using "Plugins" directory "/home/xxx/FS-UAE/Plugins"
[PLUGINS] Loading plugins from /home/xxx/FS-UAE/Plugins
fs_uae_configure_amiga_model
using PAL mode (50Hz)
[VIDEO] fs_emu_set_video_frame_rate: 50.00 (50)
[DISPLAY] Sync: g_fs_emu_video_sync_to_vblank = 0
config match for "A4000"
fs_uae_state_dir:
- checking /home/xxx/.config/fs-uae/save-states-dir
resolve_path /home/xxx/FS-UAE/Save States (absolute)
- using "Save States" directory "/home/xxx/FS-UAE/Save States"
save_states_dir: /home/xxx/FS-UAE/Save States
save_dir_name not set, using ClassicWB A4000
- using state dir /home/xxx/FS-UAE/Save States/ClassicWB A4000
final state dir path: /home/xxx/FS-UAE/Save States/ClassicWB A4000
- checking /home/xxx/.config/fs-uae/controllers-dir
resolve_path /home/xxx/FS-UAE/Controllers (absolute)
- using "Controllers" directory "/home/xxx/FS-UAE/Controllers"
- checking /home/xxx/.config/fs-uae/logs-dir
- checking /home/xxx/.config/fs-uae/cache-dir
resolve_path /home/xxx/FS-UAE/Cache (absolute)
- using "Cache" directory "/home/xxx/FS-UAE/Cache"
resolve_path /home/xxx/FS-UAE/Cache/Logs (absolute)
- using "Logs" directory "/home/xxx/FS-UAE/Cache/Logs"
switch to log file /home/xxx/FS-UAE/Cache/Logs/fs-uae.log.txt
- checking /home/xxx/.config/fs-uae/themes-dir
resolve_path /home/xxx/FS-UAE/Themes (absolute)
- using "Themes" directory "/home/xxx/FS-UAE/Themes"
themes_dir = /home/xxx/FS-UAE/Themes
fs_emu_init_2
THEME: Init
THEME: Loading ""
fs_emu_video_init
full-scene anti-aliasing is not requested
checking video sync mode
not specified: no video sync
no video sync (using timers only)
using default video format BGRA
using default texture format
fs_emu_init_render
scale: -1.00 -1.00 align: 0.50 0.50
calling fs_ml_init_2
fs_ml_input_init
calling fs_ml_video_init
using low latency vsync when full vsync is enabled
fs_ml_mouse_init
- adding system mouse
MANYMOUSE: (c) 2005-2012 Ryan C. Gordon
MANYMOUSE: Thread running
MANYMOUSE: no mice found
num joystick devices: 0
read config
fs_emu_input_init
init_input_configs
reading input device configs from /home/xxx/FS-UAE/Controllers
reading input device configs from /home/xxx/FS-UAE/Data/Devs/Keyboards
could not read directory /home/xxx/FS-UAE/Data/Devs/Keyboards
reading input device configs from /home/xxx/FS-UAE/Controllers
INPUT: Using default modifier key LALT
INPUT: Initializing devices for menu
0 Keyboard
config name "Keyboard"
get system config for device: Keyboard
checked data "input/Keyboard.conf" = 3
config file for Keyboard was not found
did not find config for device "Keyboard"
get system config for device: keyboard_0_0_0_0_linux
checked data "input/keyboard_0_0_0_0_linux.conf" = 3
config file for keyboard_0_0_0_0_linux was not found
did not find menu config for device "keyboard_0_0_0_0_linux"
get system config for device: keyboard
checked data "input/keyboard.conf" = 0
read config for keyboard for menu (from data)
checked data "input/default_keyboard.conf" = 0
read config for default_keyboard for menu (from data)
error parsing "include"
key_down (274) => action "down" (4)
did not find action for "skip_left"
key_left (276) => action "left" (1)
did not find action for "default_keyboard"
key_right (275) => action "right" (2)
key_escape (27) => action "escape" (10)
key_up (273) => action "up" (3)
did not find action for "skip_right"
key_backspace (8) => action "back" (6)
did not find action for "secondary"
key_return (13) => action "primary" (5)
1 MOUSE
config name "MOUSE"
get system config for device: MOUSE
checked data "input/MOUSE.conf" = 3
config file for MOUSE was not found
did not find config for device "MOUSE"
get system config for device: mouse_0_0_0_0_linux
checked data "input/mouse_0_0_0_0_linux.conf" = 3
config file for mouse_0_0_0_0_linux was not found
did not find menu config for device "mouse_0_0_0_0_linux"
get system config for device: mouse
checked data "input/mouse.conf" = 0
read config for mouse for menu (from data)
no config for platform menu
did not find menu config for device "mouse"
OPENAL: fs_emu_audio_init
OPENAL: openal_audio_init
[OPENAL] Opened device: OpenAL Soft
[OPENAL] Information:
[OPENAL] Devices:
- OpenAL Soft
[OPENAL] Default device: OpenAL Soft
OPENAL: trying frequency 48000
OPENAL: created context
OPENAL: made context current
openal: number of stereo sources is 2
AUDIO: Buffer target size (ms) = 40
fs_ml_video_create_window
keyboard input grab: 305419896
FS-UAE was compiled for SDL 2.0.4
[DISPLAY] 0: 1920x1080+0+0 @60
[DISPLAY] 1: 3840x2160+1920+0 @59
Monitor - j 0 pos 0
Monitor

**Comment by FrodeSolheim** [comment]
Confirmed. Not really that surprising, because FS-UAE assumes X11 on non-Windows/non-OS X platforms in order to perform some functions.

I have now a Wayland test machine running Fedora 24 w/ HD 4000 graphics, so I should be able to work on Wayland support.

**Comment by digitalstreamio** [comment]
glad to hear you can look into it. 



Thank you for looking into it. I would be glad to help out with testing if need some code tested out.

**Comment by FrodeSolheim** [comment]
@digitalstreamio Do you compile you own version? If so, I soon have some code for you to test :)

**Comment by digitalstreamio** [comment]
wow, that was fast... i roll my own builds so will grab as soon as you have
it up there.

**Comment by FrodeSolheim** [comment]
I have pushed the fixes to github :)

**Comment by digitalstreamio** [comment]
I think you nailed it. FS-UAE is now first class wayland client!

I have built the latest code and ran under F24 wayland session using SDL wayland backend. Everything looks great and is silky smooth! Video modes come up nicely, input works great. Maybe some of it is subjective, but it feels really smooth compared to X11. Thanks for putting this together. 

There is one minor issue with switching between full-screen and windowed mode. The window mode is no longer scaled after returning from full screen. It is however scaled properly when initially started  in windowed mode. Only after switching to full-screen and back to windowed, it looses the scaling.

**Comment by FrodeSolheim** [comment]
Thanks for verifying :)

About the resize issue, it looks like an SDL2 and/or Wayland issue. I added a new issue to track it here: https://github.com/FrodeSolheim/fs-uae/issues/130

**Comment by FrodeSolheim** [comment]
(Please subscribe to https://github.com/FrodeSolheim/fs-uae/issues/130 if you are interested in any updates on that, or if you get a chance to test with Fedora 25 alpha/beta, etc).
